src/prevouts.rs

```diff
@@ -1,33 +1,43 @@
 //! Revault prevouts
-//! Tiny newtype wrappers around bitcoin's OutPoint to statically check Revault transaction
-//! creation.
+//! Wrappers around bitcoin's OutPoint to statically check Revault transaction creation and ease
+//! PSBT management.
+
+use crate::txouts::{CpfpTxOut, FeeBumpTxOut, RevaultTxOut, UnvaultTxOut, VaultTxOut};
 
 use bitcoin::OutPoint;
 
 use std::fmt;
 
 /// A transaction output spent by a Revault transaction.
-pub trait RevaultPrevout: fmt::Debug + Copy {
+pub trait RevaultPrevout<T>: fmt::Debug {
     /// Get the actual outpoint
     fn outpoint(&self) -> OutPoint;
+    /// Get a reference to the txout this prevout refers to
+    fn txout(&self) -> &T;
+    /// Get the actual txout this prevout refers to
+    fn get_txout(self) -> T;
 }
 
 macro_rules! implem_revault_prevout {
     ( $struct_name:ident, $doc_comment:meta ) => {
         #[$doc_comment]
-        #[derive(Debug, Copy, Clone)]
-        pub struct $struct_name(OutPoint);
-
-        impl $struct_name {
-            /// Create a new prevout, the sequence will be set to 0xff_ff_ff_ff is None
-            pub fn new(outpoint: OutPoint) -> $struct_name {
-                $struct_name(outpoint)
-            }
+        #[derive(Debug, Clone)]
+        pub struct $struct_name<T: RevaultTxOut> {
+            outpoint: OutPoint,
+            prev_txout: T,
```

Comment:
Makes much more sense this way:
```patch
diff --git a/src/transactions.rs b/src/transactions.rs
index b9224d3..3d65127 100644
--- a/src/transactions.rs
+++ b/src/transactions.rs
@@ -377,7 +377,7 @@ impl UnvaultTransaction {
     /// addition to the unvault one.
     /// PSBT Creator and Updater.
     pub fn new(
-        vault_input: VaultTxIn<VaultTxOut>,
+        vault_input: VaultTxIn,
         unvault_txout: UnvaultTxOut,
         cpfp_txout: CpfpTxOut,
         lock_time: u32,
@@ -399,8 +399,8 @@ impl CancelTransaction {
     /// may have a fee-bumping input.
     /// PSBT Creator and Updater.
     pub fn new(
-        unvault_input: UnvaultTxIn<UnvaultTxOut>,
-        feebump_input: Option<FeeBumpTxIn<FeeBumpTxOut>>,
+        unvault_input: UnvaultTxIn,
+        feebump_input: Option<FeeBumpTxIn>,
         vault_txout: VaultTxOut,
         lock_time: u32,
     ) -> CancelTransaction {
@@ -432,8 +432,8 @@ impl EmergencyTransaction {
     /// Script. It may also spend an additional output for fee-bumping.
     /// PSBT Creator and Updater.
     pub fn new(
-        vault_input: VaultTxIn<VaultTxOut>,
-        feebump_input: Option<FeeBumpTxIn<FeeBumpTxOut>>,
+        vault_input: VaultTxIn,
+        feebump_input: Option<FeeBumpTxIn>,
         emer_txout: EmergencyTxOut,
         lock_time: u32,
     ) -> EmergencyTransaction {
@@ -465,8 +465,8 @@ impl UnvaultEmergencyTransaction {
     /// Script. It may also spend an additional output for fee-bumping.
     /// PSBT Creator and Updater.
     pub fn new(
-        unvault_input: UnvaultTxIn<UnvaultTxOut>,
-        feebump_input: Option<FeeBumpTxIn<FeeBumpTxOut>>,
+        unvault_input: UnvaultTxIn,
+        feebump_input: Option<FeeBumpTxIn>,
         emer_txout: EmergencyTxOut,
         lock_time: u32,
     ) -> UnvaultEmergencyTransaction {
@@ -499,7 +499,7 @@ impl SpendTransaction {
     /// txouts (including, but not restricted to, change).
     /// PSBT Creator and Updater.
     pub fn new(
-        unvault_inputs: Vec<UnvaultTxIn<UnvaultTxOut>>,
+        unvault_inputs: Vec<UnvaultTxIn>,
         spend_txouts: Vec<SpendTxOut>,
         lock_time: u32,
     ) -> SpendTransaction {
diff --git a/src/txins.rs b/src/txins.rs
index 9119a56..6f2caee 100644
--- a/src/txins.rs
+++ b/src/txins.rs
@@ -2,7 +2,7 @@
 //! Wrappers around bitcoin's OutPoint and previous TxOut to statically check Revault transaction
 //! creation and ease PSBT management.
 
-use crate::txouts::{CpfpTxOut, FeeBumpTxOut, RevaultTxOut, UnvaultTxOut, VaultTxOut};
+use crate::txouts::{CpfpTxOut, FeeBumpTxOut, UnvaultTxOut, VaultTxOut};
 
 use bitcoin::{OutPoint, TxIn};
 
@@ -21,25 +21,25 @@ pub trait RevaultTxIn<T>: fmt::Debug {
 }
 
 macro_rules! implem_revault_txin {
-    ( $struct_name:ident, $doc_comment:meta ) => {
+    ( $struct_name:ident, $txout_struct_name:ident, $doc_comment:meta ) => {
         #[$doc_comment]
         #[derive(Debug, Clone)]
-        pub struct $struct_name<T: RevaultTxOut> {
+        pub struct $struct_name {
             outpoint: OutPoint,
-            prev_txout: T,
+            prev_txout: $txout_struct_name,
             sequence: u32,
         }
 
-        impl<T: RevaultTxOut> RevaultTxIn<T> for $struct_name<T> {
+        impl RevaultTxIn<$txout_struct_name> for $struct_name {
             fn outpoint(&self) -> OutPoint {
                 self.outpoint
             }
 
-            fn txout(&self) -> &T {
+            fn txout(&self) -> &$txout_struct_name {
                 &self.prev_txout
             }
 
-            fn get_txout(self) -> T {
+            fn get_txout(self) -> $txout_struct_name {
                 self.prev_txout
             }
 
@@ -56,12 +56,13 @@ macro_rules! implem_revault_txin {
 
 implem_revault_txin!(
     VaultTxIn,
+    VaultTxOut,
     doc = "A vault txo spent by the unvault transaction and the emergency transaction"
 );
-impl VaultTxIn<VaultTxOut> {
+impl VaultTxIn {
     // FIXME: maybe we should take a reference to a VaultTransaction here to create the outpoint
     /// Instanciate a TxIn referencing a vault txout
-    pub fn new(outpoint: OutPoint, prev_txout: VaultTxOut, sequence: u32) -> VaultTxIn<VaultTxOut> {
+    pub fn new(outpoint: OutPoint, prev_txout: VaultTxOut, sequence: u32) -> VaultTxIn {
         VaultTxIn {
             outpoint,
             prev_txout,
@@ -72,15 +73,12 @@ impl VaultTxIn<VaultTxOut> {
 
 implem_revault_txin!(
     UnvaultTxIn,
+    UnvaultTxOut,
     doc="An unvault txo spent by the cancel transaction, an emergency transaction, and the spend transaction."
 );
-impl UnvaultTxIn<UnvaultTxOut> {
+impl UnvaultTxIn {
     /// Instanciate a TxIn referencing an unvault txout
-    pub fn new(
-        outpoint: OutPoint,
-        prev_txout: UnvaultTxOut,
-        sequence: u32,
-    ) -> UnvaultTxIn<UnvaultTxOut> {
+    pub fn new(outpoint: OutPoint, prev_txout: UnvaultTxOut, sequence: u32) -> UnvaultTxIn {
         UnvaultTxIn {
             outpoint,
             prev_txout,
@@ -91,17 +89,14 @@ impl UnvaultTxIn<UnvaultTxOut> {
 
 implem_revault_txin!(
     FeeBumpTxIn,
+    FeeBumpTxOut,
     doc = "A wallet txo spent by a revaulting (cancel, emergency) transaction to bump the transaction feerate.\
            This output is often created by a first stage transaction, but may directly be a wallet\
            utxo."
 );
-impl FeeBumpTxIn<FeeBumpTxOut> {
+impl FeeBumpTxIn {
     /// Instanciate a TxIn referencing a feebump txout
-    pub fn new(
-        outpoint: OutPoint,
-        prev_txout: FeeBumpTxOut,
-        sequence: u32,
-    ) -> FeeBumpTxIn<FeeBumpTxOut> {
+    pub fn new(outpoint: OutPoint, prev_txout: FeeBumpTxOut, sequence: u32) -> FeeBumpTxIn {
         FeeBumpTxIn {
             outpoint,
             prev_txout,
@@ -112,11 +107,12 @@ impl FeeBumpTxIn<FeeBumpTxOut> {
 
 implem_revault_txin!(
     CpfpTxIn,
+    CpfpTxOut,
     doc = "The unvault CPFP txo spent to accelerate the confirmation of the unvault transaction."
 );
-impl CpfpTxIn<CpfpTxOut> {
+impl CpfpTxIn {
     /// Instanciate a TxIn referencing a CPFP txout
-    pub fn new(outpoint: OutPoint, prev_txout: CpfpTxOut, sequence: u32) -> CpfpTxIn<CpfpTxOut> {
+    pub fn new(outpoint: OutPoint, prev_txout: CpfpTxOut, sequence: u32) -> CpfpTxIn {
         CpfpTxIn {
             outpoint,
             prev_txout,
```

Reply:
Done in last commit, can rebase it cleanly when i have a minute.